src/main/kotlin/interpreter/stack/UserCodeStackFrame.kt

```diff
@@ -329,16 +476,120 @@ class UserCodeStackFrame(val function: PyUserFunction) : StackFrame() {
+        val o1 = stack.pop()
+        val o2 = stack.pop()
+        val magic = when (type) {
+            BinaryOp.ADD -> "__add__"
+            BinaryOp.LSHIFT -> "__lshift__"
+            BinaryOp.POWER -> "__pow__"
+            BinaryOp.MULTIPLY -> "__mul__"
+            BinaryOp.MATRIX_MULTIPLY -> "__matmul__"
+            BinaryOp.FLOOR_DIVIDE -> "__floordiv__"
+            BinaryOp.TRUE_DIVIDE -> "__truediv__"
+            BinaryOp.MODULO -> "__mod__"
+            BinaryOp.SUBTRACT -> "__sub__"
+            BinaryOp.SUBSCR -> "__getitem__"
+            BinaryOp.RSHIFT -> "__rshift__"
+            BinaryOp.AND -> "__and__"
+            BinaryOp.XOR -> "__xor__"
+            BinaryOp.OR -> "__or__"
+            else -> error("This should never happen!")
+        }
+        magicMethod(o1, magic, o2, "__r" + magic.substring(2))
+        bytecodePointer += 1
+    }
+
+    /**
+     * INPLACE_*
+     */
+    fun inplaceOp(type: BinaryOp, arg: Byte) {
+        val o1 = stack.pop()
+        val o2 = stack.pop()
+        val magic = when (type) {
+            BinaryOp.ADD -> "__iadd__"
+            BinaryOp.LSHIFT -> "__ilshift__"
+            BinaryOp.POWER -> "__ipow__"
+            BinaryOp.MULTIPLY -> "__imul__"
+            BinaryOp.MATRIX_MULTIPLY -> "__imatmul__"
+            BinaryOp.FLOOR_DIVIDE -> "__ifloordiv__"
+            BinaryOp.TRUE_DIVIDE -> "__itruediv__"
+            BinaryOp.MODULO -> "__imod__"
+            BinaryOp.SUBTRACT -> "__isub__"
+            BinaryOp.SUBSCR -> "__igetitem__"
+            BinaryOp.RSHIFT -> "__irshift__"
+            BinaryOp.AND -> "__iand__"
+            BinaryOp.XOR -> "__ixor__"
+            BinaryOp.OR -> "__ior__"
+            BinaryOp.STORE_SUBSCR -> "__setitem__"
+            BinaryOp.DELETE_SUBSCR -> "__delitem__"
+        }
+        magicMethod(o1, magic, o2)
+        bytecodePointer += 1
+    }
+
+    fun compareOp(arg: Byte) {
+        val top = stack.pop()
+        val second = stack.pop()
+        when (arg.toInt()) {
+            0 -> magicMethod(top, "__lt__", second, "__ge__")
+            1 -> magicMethod(top, "__le__", second, "__gt__")
+            2 -> magicMethod(top, "__gt__", second, "__le__")
+            3 -> magicMethod(top, "__ge__", second, "__lt__")
+            4 -> magicMethod(top, "__eq__", second, "__eq__")
+            5 -> magicMethod(top, "__ne__", second, "__ne__")
+            6 -> magicMethod(top, "__contains__", second)
+            7 -> {
+                magicMethod(top, "__contains__", second)
+                stack.push(if (stack.pop() == PyBool.TRUE) PyBool.FALSE else PyBool.TRUE)
             }
-            // Kept for future-proofing.
-            else -> TODO("Unsupported binary op $type")
+            8 -> stack.push(if (top.hashCode() == second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            9 -> stack.push(if (top.hashCode() != second.hashCode()) PyBool.TRUE else PyBool.FALSE)
+            10 -> TODO("exception match COMPARE_OP")
```

Comment:
Else branch

Reply:
Else branch should be an error(), because it usually means we didn't handle a param.